java/src/jmri/jmrix/dccpp/DCCppProgrammer.java

```diff
@@ -264,9 +264,7 @@ synchronized public void message(DCCppMessage l) {
     // Handle a timeout notification
     @Override
     public void notifyTimeout(DCCppMessage msg) {
-        if (log.isDebugEnabled()) {
-            log.debug("Notified of timeout on message{}", msg.toString());
-        }
+        log.debug("Notified of timeout on message '{}'", msg.toString());
```

Comment:
Is `toString()` required here? I'm not sure, but I think you can remove the `toString()` call.
Maybe the toString() call is cheap and if so, not a problem. But the reason for the check of `log.isDebugEnabled()` is to not call toString() if debug is disabled.

If you have this line as below instead, this problem goes away:
`log.debug("Notified of timeout on message '{}'", msg);`

I think this will work since I think log.debug will call msg.toString(), but I'm not sure and I haven't tested it lately.

Reply:
My understanding is that the arguments are not evaluated unless the log statement is enabled. This is documented here: http://www.slf4j.org/faq.html#logging_performance

Reply:
The arguments are evaluated as part of collecting for the function call.  That's inevitable given how the JVM works.

What's not evaluated is the creation of the final logging string.  That includes inserting the arguments into the string, and _that_ includes converting the arguments to strings as needed.

So if you put `foo.toString()` as an argument in the function call, that `.toString()` will be evaluated at every call, whether or not the statement actually logs.  If you just put `foo` as the argument, the `.toString()` is invoked only when the statement is actually going to log.

Reply:
OK, thanks for that explanation. I missed that distinction.